### PR TITLE
Improve peformance of CTransaction::HasWitness (28107 follow-up)

### PR DESCRIPTION
**Pull request by dergoegge**
This addresses https://github.com/bitcoin/bitcoin/pull/28107#discussion_r1364961590 from #28107.

### CODE REVIEW

**Comment by dergoegge** [comment]
Can someone please re-run the "CI / macOS 13 native" job, failure seems unrelated.

**Comment by fanquake** [comment]
>  failure seems unrelated.

Are you sure?
```bash
 Traceback (most recent call last):
  File "/Users/runner/work/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/Users/runner/work/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin/test/functional/rpc_createmultisig.py", line 65, in run_test
    self.do_multisig()
  File "/Users/runner/work/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin/test/functional/rpc_createmultisig.py", line 241, in do_multisig
    tx = node0.sendrawtransaction(rawtx3["hex"], 0)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin/test/functional/test_framework/coverage.py", line 50, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/bitcoin/bitcoin/ci/scratch/build/bitcoin-x86_64-apple-darwin/test/functional/test_framework/authproxy.py", line 129, in __call__
    raise JSONRPCException(response['error'], status)
test_framework.authproxy.JSONRPCException: mandatory-script-verify-flag-failed (Witness program was passed an empty witness) (-26)
```

**Comment by dergoegge** [comment]
> Are you sure?

Yes, it passes locally.

**Comment by fanquake** [comment]
Want to document it as an intermittent failure?

**Review by stickies-v — COMMENTED** [review]
Approach ACK

I had similar thoughts while reviewing 28107 but figured the computation was still fast enough. Given that `ComputeWitnessHash` is only called once during initialization (and is private) and `HasWitness` is called frequently, this approach is strictly better.

**Comment by fanquake** [comment]
The `rpc_createmultisig.py` test is now failing in two other CI jobs and MSAN is unhappy:
```bash
==19201==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x563f069e85d0 in bcmp /msan/llvm-project/compiler-rt/lib/msan/../sanitizer_common/sanitizer_common_interceptors.inc:856:10
    #1 0x563f0956bff5 in base_blob<256u>::Compare(base_blob<256u> const&) const src/./uint256.h:54:66
    #2 0x563f0956bff5 in operator!=(base_blob<256u> const&, base_blob<256u> const&) src/./uint256.h:57:89
    #3 0x563f0956bff5 in CTransaction::HasWitness() const src/./primitives/transaction.h:370:33
    #4 0x563f0956bff5 in void SerializeTransaction<CHashWriter, CTransaction>(CTransaction const&, CHashWriter&) src/./primitives/transaction.h:265:16
    #5 0x563f09569863 in void CTransaction::Serialize<CHashWriter>(CHashWriter&) const src/./primitives/transaction.h:326:9
    #6 0x563f09569863 in void Serialize<CHashWriter, CTransaction>(CHashWriter&, CTransaction const&) src/./serialize.h:776:7
    #7 0x563f09569863 in CHashWriter& CHashWriter::operator<<<CTransaction>(CTransaction const&) src/./hash.h:161:9
    #8 0x563f09569863 in CTransaction::ComputeWitnessHash() const src/primitives/transaction.cpp:93:47
    #9 0x563f09569f47 in CTransaction::CTransaction(CMutableTransaction&&) src/primitives/transaction.cpp:97:190
    #10 0x563f06eb3f68 in std::__1::__shared_ptr_emplace<CTransaction const, std::__1::allocator<CTransaction const>>::__shared_ptr_emplace[abi:v170002]<CMutableTransaction>(std::__1::allocator<CTransaction const>, CMutableTransaction&&) /msan/cxx_build/include/c++/v1/__memory/shared_ptr.h:302:37
    #11 0x563f06eb3f68 in std::__1::shared_ptr<CTransaction const> std::__1::allocate_shared[abi:v170002]<CTransaction const, std::__1::allocator<CTransaction const>, CMutableTransaction, void>(std::__1::allocator<CTransaction const> const&, CMutableTransaction&&) /msan/cxx_build/include/c++/v1/__memory/shared_ptr.h:1022:55
    #12 0x563f09071b26 in std::__1::shared_ptr<CTransaction const> std::__1::make_shared[abi:v170002]<CTransaction const, CMutableTransaction, void>(CMutableTransaction&&) /msan/cxx_build/include/c++/v1/__memory/shared_ptr.h:1031:12
    #13 0x563f09071b26 in std::__1::shared_ptr<CTransaction const> MakeTransactionRef<CMutableTransaction>(CMutableTransaction&&) src/./primitives/transaction.h:418:93
    #14 0x563f09071b26 in ChainstateManager::UpdateUncommittedBlockStructures(CBlock&, CBlockIndex const*) const src/validation.cpp:3706:24
    #15 0x563f09073e46 in ChainstateManager::GenerateCoinbaseCommitment(CBlock&, CBlockIndex const*) const src/validation.cpp:3733:5
    #16 0x563f08bb9dd6 in node::BlockAssembler::CreateNewBlock(CScript const&) src/node/miner.cpp:160:69
    #17 0x563f0822b2a5 in TestChain100Setup::CreateBlock(std::__1::vector<CMutableTransaction, std::__1::allocator<CMutableTransaction>> const&, CScript const&, Chainstate&) src/test/util/setup_common.cpp:310:56
```

**Review by glozow — COMMENTED** [review]
Concept ACK

File: src/primitives/transaction.h [hunk 1]
```diff
@@ -367,12 +367,7 @@ class CTransaction
 
     bool HasWitness() const
     {
-        for (size_t i = 0; i < vin.size(); i++) {
-            if (!vin[i].scriptWitness.IsNull()) {
-                return true;
-            }
-        }
-        return false;
+        return hash.ToUint256() != m_witness_hash.ToUint256();
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/primitives/transaction.h:370]: Making them optional would add 16 bytes to every `CTransaction` struct,  I think? This approach would also prevent them from being marked `const` in `CTransaction`. If having `m_witness_hash` be non-const was okay, you could default it to `uint256::ZERO` and set `m_witness_hash = ComputeWitnessHash()` in the constructor. I don't think that's ideal though.

**Comment by ajtowns** [comment]
> ```shell
>  #3 0x563f0956bff5 in CTransaction::HasWitness() const src/./primitives/transaction.h:370:33
>  #4 0x563f0956bff5 in void SerializeTransaction<CHashWriter, CTransaction>(CTransaction const&, CHashWriter&) src/./primitives/transaction.h:265:16
> ```

If it's not obvious, the bug is that `m_witness_hash` is initialised by serializing the transaction with its witness, but in that case, `Serialize` calls `HasWitness()` to work out which format to use, which with this change relies on access to `m_witness_hash` which is not yet initialised.

**Comment by dergoegge** [comment]
Added a `std::optional<bool>` to `CTransaction` as a cache for `HasWitness`.

**Comment by stickies-v** [comment]
I'm not a fan of the `std::optional<bool>` cache approach, it's an error-prone API imo (even if it's private and small). I think it makes more sense to just calculate the boolean at initialization and follow the same pattern as for `hash` and `m_witness_hash`? It has to be calculated for every instance anyway.

As e.g. in https://github.com/stickies-v/bitcoin/commit/2e69a6597c1a07b319e9cba4601474fbe0761f71

**Comment by dergoegge** [comment]
Thanks @stickies-v went with your suggestion as my previous idea was buggy again🔥

File: src/primitives/transaction.h [hunk 2]
```diff
@@ -310,12 +310,15 @@ class CTransaction
 
 private:
     /** Memory only. */
+    const bool m_has_witness;
```
Comment (ajtowns) [inline src/primitives/transaction.h:313]: Not really a fan of adding 8 bytes to every `CTransaction`, though I guess it's not horrible?

Another approach would be to separate out the raw data and compute-logic into one class, and the optimised version with the w/txid cached into a separate class. That looks like: https://github.com/ajtowns/bitcoin/commit/89e3c808927984309a002b2cab1c0a622f5c15f8

Comment (dergoegge) [reply src/primitives/transaction.h:313]: Personally, I don't think the extra bytes matter that much, so I would lean towards leaving this as is.

**Comment by theStack** [comment]
Concept ACK

**Review by theStack — COMMENTED** [review]
The code looks correct.

I'm still unsure on how to assess whether the extra memory needed is okay or not, other than "gut feeling". I would have expected that the increased size of `CTransaction` (on my system, it's 128 bytes on the PR vs. 120 bytes on master, a growth of ~6.66%) has a tiny impact on the mempool usage accounting as well, but it seems that it doesn't, as `RecursiveDynamicUsage` only cares about a tx's inputs and outputs fields:
https://github.com/bitcoin/bitcoin/blob/8243762700bc6e7876ae5d4fc000500858b99f66/src/core_memusage.h#L32-L41

(A bit off-topic to this PR, but is it really intended to _only_ do dynamic accounting? Just because something doesn't grow dynamically doesn't mean it's completely free, the static parts still have to reside somewhere in memory.)

**Comment by theStack** [comment]
@sipa: Thanks for clarifying, that makes sense and I see now why it would be a bad idea to add the static part directly in the `DynamicMemoryUsage` functions. For the specific case of mempool usage accounting, one could add `sizeof(CTransaction)` at the call-site though? E.g.
```diff
diff --git a/src/kernel/mempool_entry.h b/src/kernel/mempool_entry.h
index 1f175a5ccf..95362fe2d9 100644
--- a/src/kernel/mempool_entry.h
+++ b/src/kernel/mempool_entry.h
@@ -108,7 +108,7 @@ public:
         : tx{tx},
           nFee{fee},
           nTxWeight{GetTransactionWeight(*tx)},
-          nUsageSize{RecursiveDynamicUsage(tx)},
+          nUsageSize{sizeof(*tx) + RecursiveDynamicUsage(tx)},
           nTime{time},
           entry_sequence{entry_sequence},
           entryHeight{entry_height},
```
(Not saying it's really worth it to fiddle around in this area, and I personally wouldn't feel comfortable to open a PR. I was just very surprised that once could -- in theory -- bloat up `CTransaction` and it still goes unnoticed for the mempool usage accounting.)

**Comment by theStack** [comment]
@sipa: Ah, very interesting, have to admit I completely overlooked `RecursiveDynamicUsage(const std::shared_ptr<X>& p)`.

> Is it possible that you just don't see the added field cause changes because either it's fitting in earlier padding space, or because the rounding up effect of `malloc()` hides the difference?

Yes, it was the latter. The added field did lead to a size increase of `CTransaction` (from 120 to 120 bytes), but `MallocUsage` indeed yields the same value for both of these [sizes on a 64-bit machine](https://github.com/bitcoin/bitcoin/blob/8243762700bc6e7876ae5d4fc000500858b99f66/src/memusage.h#L57) (namely ((120+31)>>4)<<4 = ((128+31)>>4)<<4 = 144). So that's the reason why I did see the same mempool usage accounting results both on master and the PR on my machine (and wrongly concluded from that that the size of `CTransaction` is not accounted for). To be sure, I intentionally bloated `CTransaction` up much more with an `const uint8_t dummy[100] = {};` entry and verified that the mempool usage was higher indeed. So, the accounting seems to work as expected and my worries were unjustified.

If anyone wants to run these experiments: what I did was to start bitcoind with an existing mempool.dat in datadir both on master and the PR branch with `-nolisten -noconnect` (to avoid clogging the mempool with new txs), wait until the mempool loaded, ran `bitcoin-cli getmempoolinfo` and compared the `"usage"` result field value (other fields like `"size"`, `"bytes"` and `"total_fee"` should always be equal for the same mempool.dat).

ACK af1d2ff88344e1545ac8d9ad09f8e37e264da712

(still based on a gut feeling, but garnished with some napkin math: the mempool on my node is close to it's maximum of 300MB and has about ~100k entries now; `MallocUsage` increases in 16-bytes steps on 64-bit machines, so that would in the worst case lead to ~1,6MB more usage, where now "only" ~99.5k entries would fit; I think that's fine).

File: src/primitives/transaction.h (hunk 2 above)
Comment (stickies-v) [inline src/primitives/transaction.h:313]: nit: Using default initializer makes more sense here I think:

<details>
<summary>git diff on af1d2ff883</summary>

```diff
diff --git a/src/primitives/transaction.cpp b/src/primitives/transaction.cpp
index 77a363f7b6..a20cf25f53 100644
--- a/src/primitives/transaction.cpp
+++ b/src/primitives/transaction.cpp
@@ -93,8 +93,10 @@ Wtxid CTransaction::ComputeWitnessHash() const
     return Wtxid::FromUint256((CHashWriter{0} << *this).GetHash());
 }
 
-CTransaction::CTransaction(const CMutableTransaction& tx) : vin(tx.vin), vout(tx.vout), nVersion(tx.nVersion), nLockTime(tx.nLockTime), m_has_witness{ComputeHasWitness()}, hash{ComputeHash()}, m_witness_hash{ComputeWitnessHash()} {}
-CTransaction::CTransaction(CMutableTransaction&& tx) : vin(std::move(tx.vin)), vout(std::move(tx.vout)), nVersion(tx.nVersion), nLockTime(tx.nLockTime), m_has_witness{ComputeHasWitness()}, hash{ComputeHash()}, m_witness_hash{ComputeWitnessHash()} {}
+CTransaction::CTransaction(const CMutableTransaction& tx)
+    : vin(tx.vin), vout(tx.vout), nVersion(tx.nVersion), nLockTime(tx.nLockTime) {}
+CTransaction::CTransaction(CMutableTransaction&& tx)
+    : vin(std::move(tx.vin)), vout(std::move(tx.vout)), nVersion(tx.nVersion), nLockTime(tx.nLockTime) {}
 
 CAmount CTransaction::GetValueOut() const
 {
diff --git a/src/primitives/transaction.h b/src/primitives/transaction.h
index 594168bbcc..facbd8c769 100644
--- a/src/primitives/transaction.h
+++ b/src/primitives/transaction.h
@@ -310,9 +310,9 @@ public:
 
 private:
     /** Memory only. */
-    const bool m_has_witness;
-    const Txid hash;
-    const Wtxid m_witness_hash;
+    const bool m_has_witness{ComputeHasWitness()};
+    const Txid hash{ComputeHash()};
+    const Wtxid m_witness_hash{ComputeWitnessHash()};
 
     Txid ComputeHash() const;
     Wtxid ComputeWitnessHash() const;

```
</details>

**Review by stickies-v — APPROVED** [review]
ACK af1d2ff88344e1545ac8d9ad09f8e37e264da712

I think the PR and commit description would benefit from describing that this is not a strict improvement, as we slightly increase memory usage (which I think is an acceptable trade-off).

File: src/primitives/transaction.h (hunk 2 above)
Comment (ajtowns) [reply src/primitives/transaction.h:313]: Really, we're already wasting 16 bytes due to `vin` and `vout` supporting `capacity() != size()` despite both being `const` and never needing to be resized.

**Comment by dergoegge** [comment]
I think this is rfm. Could a maintainer take a look?

**Review by sedited — APPROVED** [review]
ACK af1d2ff88344e1545ac8d9ad09f8e37e264da712

Nit: Maybe add a `HasWitness` check somewhere in the `transaction_tests`?

**Comment by mzumsande** [comment]
Since it comes with a memory cost, I wonder if the the performance gain we get for it can be measured. Does this speed up any benchmarks / IBD?

**Comment by dergoegge** [comment]
> Since it comes with a memory cost, I wonder if the the performance gain we get for it can be measured. Does this speed up any benchmarks / IBD?

It's less of a gain and more just trying to revert the performance regression I introduced in #28017. I guess you are asking if it is a noticeable regression. I would think both the performance regression and the new memory cost from here are negligible and I can certainly also live with what is in master now.

**Review by mzumsande — COMMENTED** [review]
> I guess you are asking if it is a noticeable regression. I would think both the performance regression and the new memory cost from here are negligible and I can certainly also live with what is in master now.

I think this should be easily benched just with a `--reindex-chainstate`? I could try that next week if no one else wants to. I just think that if it's possible within reasonable efforts, this kind of decisions should be made based on empirical data rather than guessing what the performance / memory impacts may be.

**Comment by dergoegge** [comment]
> this kind of decisions should be made based on empirical data rather than guessing what the performance / memory impacts may be.

This makes it sound like this is all up in the air when the trade-off for this PR is well known. All this does is revert to the performance we had pre-#28017 while adding an extra 8 bytes to every CTransaction instance.

The runtime complexity of `HasWitness` is now `O(tx.vin.size())` while it used to be `O(1)`. That could be considered good enough reasoning for this PR even if a IBD/reindex benchmark doesn't show a noticeable difference. The 4 concept acks and 3 full acks seem to agree on that.

In any case the memory usage of CTransaction could be optimized in a follow-up, even beyond reverting the overhead from this PR (https://github.com/bitcoin/bitcoin/pull/28766#discussion_r1394031188).

Feel free to do the benchmark, but I'm not sure if the result matters much at this point, since we have already spend sufficient review time on this PR.

**Comment by dergoegge** [comment]
> All this does is revert to the performance we had pre-https://github.com/bitcoin/bitcoin/issues/28017 while adding an extra > 8 bytes to every CTransaction instance.
>
> The runtime complexity of HasWitness is now O(tx.vin.size()) while it used to be O(1).

Whoops, this isn't really true. HasWitness was unchanged by #28107, the only change in this regard was https://github.com/bitcoin/bitcoin/pull/28107/commits/cdb14d79e809bf7d1612b21b554a9fcfb2ab1c91.

This PR improves the performance for all call sites of HasWitness not just the ones in https://github.com/bitcoin/bitcoin/pull/28107/commits/cdb14d79e809bf7d1612b21b554a9fcfb2ab1c91.

**Comment by achow101** [comment]
ACK af1d2ff88344e1545ac8d9ad09f8e37e264da712

**Pull request merged**